### PR TITLE
Proj 6 coordinate operation handling -- report errors and warnings to users

### PR DESCRIPTION
**Pull request by nyalldawson**
This PR extends the Proj 6 support available in master with methods to allow warnings and errors from proj to be pushed back and displayed to users. This is critical functionality -- by hiding these warnings and errors we allow users to accidentally create nonsensical or invalid coordinate transformations silently. Better to push the warnings into their faces so that they can make informed decisions based on this.

In particular,  warnings are shown when:

- A coordinate transform cannot be created, e.g. due to not making sense (transforming earth coords to mars coords)
- A coordinate transform cannot be created due to a missing grid file (along with links to download the grid if available)
- A coordinate transform was created, BUT a better one is possible, it just is missing some prerequisite on the user's system (again, with links to download if available)

Sponsored by ICSM

### CODE REVIEW

**Comment by nyalldawson** [comment]
Screencast:

Peek 2019-05-31 13-55

**Comment by nirvn** [comment]
@nyalldawson , nice. Out of curiosity from looking at the GIF, what does "an accuracy of -1 [i.e. negative one] meter" means?

**Comment by nyalldawson** [comment]
@nirvn fixed -- it means "unknown"

**Comment by wonder-sk** [comment]
wow - this is really great

**Comment by nyalldawson** [comment]
IMO this completes the epic QGIS -> proj 6 upgrade. May I well and truly be retired before the next one ;)

**Comment by nirvn** [comment]
:tada:

File: src/core/qgscoordinatetransform.cpp [hunk 1]
```diff
@@ -652,7 +652,18 @@ void QgsCoordinateTransform::transformCoords( int numPoints, double *x, double *
...
+  int projResult = 0;
+#if PROJ_VERSION_MAJOR>=6
+  proj_errno_reset( projData );
+  proj_trans_generic( projData, direction == ForwardTransform ? PJ_FWD : PJ_INV,
+                      x, sizeof( double ), numPoints,
+                      y, sizeof( double ), numPoints,
+                      z, sizeof( double ), numPoints,
+                      nullptr, sizeof( double ), 0 );
+  projResult = proj_errno( projData );
+  // ewww - this logic is gross. We should drop support for PROJ 6.0 as quickly as possible and dump this code (in favour of built in methods used for >=6.1 builds)
```
Comment (nirvn) [inline src/core/qgscoordinatetransform.cpp]: I believe you can get rid of the eww comment here.

Comment (nyalldawson) [reply src/core/qgscoordinatetransform.cpp]: Thanks, fixed

**Review by nirvn — APPROVED** [review]

**Comment by nyalldawson** [comment]
Woohoo!

**Comment by rduivenvoorde** [comment]
Thanks @nyalldawson  !!

**Comment by PedroVenancio** [comment]
Really really great stuff @nyalldawson !!!

Thank you very much to you, ICSM and PROJ/GDAL guys!!!

Can the warn message show where the grid shift files should be manually pasted (proj folder), when they are missing? This can be useful for users.

**Pull request merged**